### PR TITLE
session, tikv: allocate task IDs for distsql requests

### PR DESCRIPTION
**Pull request by sticnarf**
- [x] Depends on https://github.com/pingcap/kvproto/pull/604

### What problem does this PR solve?

Issue Number: close https://github.com/tikv/tikv/issues/7484

Problem Summary:

Currently TiDB provides no information for TiKV that some group of requests belong to the same statement (task). So TiKV's unified read pool uses the start TS of the transaction as the task identifier . It's not suitable in all cases to use start ts as the task id for the unified read pool. For example, in the same transaction, a big query is executed and the following point selects are also downgraded.

### What is changed and how it works?

What's Changed:

Currently in this PR, we only set task ids for distsql requests that can be run in transactions. For other distsql requests (like DDL reorg, analyze and checksum), the task id is not provided and TiKV still uses start_ts as task id. And TiKV uses random numbers as the task ids for point gets commands so this PR also does not set task ids for these requests.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

### Release note 

Allocate task IDs for distsql requests to help TiKV schedule them better.

### CODE REVIEW

**Comment by breeswish** [comment]
Should it be called SqlStatementId instead?

File: sessionctx/stmtctx/stmtctx.go [hunk 1]
```diff
@@ -38,6 +39,13 @@ const (
 	WarnLevelNote = "Note"
 )
 
+var taskIDAlloc uint64
+
+// AllocateTaskID allocates a new unique ID for a statement execution
+func AllocateTaskID() uint64 {
+	return atomic.AddUint64(&taskIDAlloc, 1)
```
Comment (zz-jason) [inline sessionctx/stmtctx/stmtctx.go:46]: There is a chance that different tidb-server use the same task ID.

Comment (BusyJay) [reply sessionctx/stmtctx/stmtctx.go:46]: TiKV will combine both `start_ts` and `task_id` into one.

See https://github.com/tikv/tikv/pull/7521/files#diff-d674d3a094c25d9060532e4d2ff339d1R158-R178

File: sessionctx/stmtctx/stmtctx.go [hunk 2]
```diff
@@ -38,6 +39,13 @@ const (
 	WarnLevelNote = "Note"
 )
 
+var taskIDAlloc uint64
```
Comment (XuHuaiyu) [inline sessionctx/stmtctx/stmtctx.go:42]: taskIDAlloc is a server-level variable here.
Can we define it as a session-level variable to reduce the contention on it?

File: sessionctx/stmtctx/stmtctx.go (hunk 1 above)
Comment (zz-jason) [reply sessionctx/stmtctx/stmtctx.go:46]: Got it, how about moving this calculation logic into the request sender side?

Comment (BusyJay) [reply sessionctx/stmtctx/stmtctx.go:46]: TiDB's logic is too complicated, I'm not sure if start ts exists when task id is assigned.

Comment (coocood) [reply sessionctx/stmtctx/stmtctx.go:46]: Can we use the statement idx, starts from 0 as the task id in the transaction?

Comment (BusyJay) [reply sessionctx/stmtctx/stmtctx.go:46]: The original author also suggests to use the ID for tracing. I'm not sure if allocating as statement index will affect the purpose.

**Comment by sticnarf** [comment]
PTAL again @XuHuaiyu @zz-jason 
Thanks!

**Comment by SunRunAway** [comment]
Does this PR need a release note? It seems not to affect users

**Comment by sticnarf** [comment]
> Does this PR need a release note? It seems not to affect users

Currently the only effect is to help TiKV schedule requests better. I think you can decide whether to include the release note as you please.

**Pull request merged**